### PR TITLE
Fix GDScriptCompiler bugging out with identically-named inner class

### LINKED ISSUE #27111: `is` returns true when inner class names match.
**Godot version:**

3.1 stable

**OS/device including version:**

Mac OS Sierra 10.12.6

**Issue description:**

The `is` function is returning `true` when an instance has an Inner Class name that matches what it is being compared to even if the Outer Class names do not match.

**Steps to reproduce:**
The following script will print "I think we have a bug here."
```
extends SceneTree

class HasInnerClass1:
	class InnerClass:
		var a = 1

class HasInnerClass2:
	class InnerClass:
		var a = 2

func _init():
	var inner_of_hasInnerClass1 = HasInnerClass1.InnerClass.new()
	if(inner_of_hasInnerClass1 is HasInnerClass2.InnerClass):
		print('I think we have a bug here.')
	else:
		print('Everything checks out.')

	quit()
```

**Minimal reproduction project:**

See script in Steps to Reproduce

### PR DESCRIPTION
**Pull request by bojidar-bg**
Fixes #27111.

### CODE REVIEW

**Comment by akien-mga** [comment]
> 09:55 <bojidar_bg> Is there someone around with large GDScript projects which use inner classes? Would be nice if you can try running your project with #27128 (make backups, just in case!)

@avencherus maybe?

File: modules/gdscript/gdscript_compiler.h [hunk 1]
```diff
@@ -38,9 +38,8 @@
 class GDScriptCompiler {
 
 	const GDScriptParser *parser;
-	Map<StringName, Ref<GDScript> > class_map;
-	Set<StringName> parsed_classes;
-	Set<StringName> parsing_classes;
+	Set<GDScript*> parsed_classes;
+	Set<GDScript*> parsing_classes;
```
Comment (akien-mga) [inline modules/gdscript/gdscript_compiler.h]: Style issue: https://travis-ci.org/godotengine/godot/jobs/507240380

Comment (bojidar-bg) [reply modules/gdscript/gdscript_compiler.h]: Hm, seems like the clang-format plugin wasn't configured to work on .c/.h files. Thanks!

**Comment by avencherus** [comment]
@akien-mga No longer unfortunately.  Months ago I had to drop many of these features, and reverted to previous Godot coding styles, because various issues with class names were shutting me down completely.

When there is time for small projects I will periodically test the waters, but it is going to be quite awhile before we start our next big project though.  X)

**Comment by vnen** [comment]
Looks good to me, but I guess it needs extensive testing.

**Comment by akien-mga** [comment]
Thanks!

> Looks good to me, but I guess it needs extensive testing.

Agreed, let's wait for it to be tested in the master branch before considering a cherry-pick @hpvb (still keeping the label so that we don't forget about it).

**Comment by akien-mga** [comment]
If cherry-picked, #27867 should go with it.

**Comment by hpvb** [comment]
Consider for 3.1.2

**Pull request merged**